servant/src/Servant/API/Cookie.hs

```diff
@@ -0,0 +1,23 @@
+{-# LANGUAGE DataKinds          #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DeriveFunctor      #-}
+{-# LANGUAGE PolyKinds          #-}
+{-# OPTIONS_HADDOCK not-home    #-}
+module Servant.API.Cookie where
+
+import Data.ByteString    (ByteString)
+import Data.Typeable      (Typeable)
+import GHC.TypeLits       (Symbol)
+
+-- | Extract the given cookie's value as a value of type @a@.
+--
+-- Example:
+--
+-- >>>            -- GET /test
+-- >>> type MyApi = "test" :> Cookie "token" Text :> Get '[JSON] Text
+
+data Cookie (sym :: Symbol) a = Cookie a
+                              | MissingCookie
```

Comment:
No, I just copied `Header` and didn't know if I would need them at the time. I'll see if they come up when I implement `HasClient` etc and if not remove them.